Tests/TensorFlowTests/EpochsTests.swift

```diff
@@ -19,95 +19,156 @@ import XCTest
 var rng = ARC4RandomNumberGenerator(seed: [42])
 
 final class EpochsTests: XCTestCase {
-  func testBaseUse() {
-    // A mock item type that tracks if it was accessed or not
-    class AccessTracker {
+ 
+  // An element that keeps track of when it was first accessed.
+  class AccessTracker {
       var accessed: Bool = false
+  }
+  
+  // A struct keeping track of when its elements have been first accessed. We 
+  // use it in the tests to check whether methods that are not supposed to break
+  // the laziness work as intended.
+  struct ReadTracker<Base: RandomAccessCollection> : RandomAccessCollection {
+    let base: Base
+    let accessed_: [AccessTracker]
+      
+    public typealias Element = Base.Element
+    /// A type whose instances represent positions in `self`.
+    public typealias Index = Base.Index
+    /// The position of the first element.
+    public var startIndex: Index { base.startIndex }
+    /// The position one past the last element.
+    public var endIndex: Index { base.endIndex }
+    /// Returns the position after `i`.
+    public func index(after i: Index) -> Index { base.index(after: i) }
+    /// Returns the position after `i`.
+    public func index(before i: Index) -> Index { base.index(before: i) }
+
+    init(_ base: Base) { 
+      self.base = base
+      accessed_ = (0..<base.count).map { _ in AccessTracker() } 
+    }  
+    
+    subscript(i: Base.Index) -> Base.Element {
```

Comment:
Doc comment missing.  It might be good to repeat the warning about thread safety here.

Reply:
Addressed.